merlion/models/layers.py

```diff
@@ -0,0 +1,253 @@
+        return config_dict
+
+    def __copy__(self):
+        config_dict = super().to_dict(_skipped_keys={"model"})
+        return self.__class__(model=deepcopy(self.model), **config_dict)
+
+    def __getattr__(self, item):
+        base_model = self.base_model
+        is_detector_attr = isinstance(base_model, DetectorBase) and hasattr(DetectorConfig, item)
+        is_forecaster_attr = isinstance(base_model, ForecasterBase) and hasattr(ForecasterConfig, item)
+        if is_detector_attr or is_forecaster_attr:
+            return getattr(base_model.config, item)
+        return self.__getattribute__(item)
+
+
+class LayeredModel(ModelBase, metaclass=AutodocABCMeta):
+    """
+    Abstract class implementing a model which wraps around another internal model.
+
+    The actual underlying model is stored in ``model.config.model``, and ``model.model`` is a property which references
+    this. This is to allow the model to retain the initializer ``LayeredModel(config)``, and to ensure that various
+    attributes do not become de-synchronized (e.g. if we were to store ``config.model_config`` and ``model.model``
+    separately).
+
+    We define the _base model_ as the non-layered model at the base of the overall model hierarchy.
+
+    The layered model is allowed to access any callable attribute of the base model,
+    e.g. ``model.set_seasonality(...)`` resolves to``model.base_model.set_seasonality(...)`` for a `SeasonalityModel`.
+    If the base model is a forecaster, the layered model will automatically inherit from `ForecasterBase`; similarly
+    for `DetectorBase` or `ForecastingDetectorBase`. The abstract methods (``forecast`` and ``get_anomaly_score``)
+    are overridden to call the underlying model.
+
+    If the base model is a forecaster, the top-level config ``model.config`` does not duplicate attributes of the
+    underlying forecaster config (e.g. ``max_forecast_steps`` or ``target_seq_index``). Instead,
+    ``model.config.max_forecast_steps`` will resolve to ``model.config.base_model.max_forecast_steps``.
+    As a result, you will only need to specify this parameter once. The same holds true for `DetectorConfig` attributes
+    (e.g. ``threshold`` or ``calibrator``) when the base model is an anomaly detector.
+
+    .. note::
+
+        For the time being, every layer of the model is allowed to have its own ``transform``.
+    """
+
+    config_class = LayeredModelConfig
+    require_even_sampling = False
+    require_univariate = False
+
+    def __new__(cls, config: LayeredModelConfig = None, model: ModelBase = None, **kwargs):
+        # Dynamically inherit from the appropriate kind of base model
+        original_cls = cls
+        config = cls._resolve_args(config=config, model=model, **kwargs)
+        if isinstance(config.model, ForecastingDetectorBase):
+            cls = cls.__class__(cls.__name__, (cls, LayeredForecastingDetector), {})
+            setattr(cls, "_original_cls", original_cls)
```

Comment:
`_original_cls` seems like a crucial piece of the Layer code, can we document it somewhere in this class?

Reply:
Added the following explanation:
```python
# Dynamically inherit from the appropriate kind of base model.
# However, this creates a new class that isn't registered anywhere with pickle/dill. This causes
# serialization problems, especially when using models with multiprocessing. So we maintain this
# class (cls) as a class attribute _original_cls of the new, dynamically created class. This is used 
# by the __reduce__ method when pickling a LayeredModel.
```